### PR TITLE
[Refactor] Move status query into the cloud class

### PR DESCRIPTION
**Pull request by Michaelvll**
Previously, the cloud-specific status query functions are located in the `backend_utils` making adding new clouds and features for clouds not clean.

We now move the status query to the cloud class, making each cloud more modularized.

This PR is to simplify the process of adding cluster migration #2088.

This PR also fixes the status refresh for Azure multi-node cluster! Previously, refreshing the status for the multi-node Azure cluster will remove the cluster from table immediately.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Manual tests:
  - [x] `sky launch -c min --cloud aws --cpus 2 echo hi -i 0; sky status -r`
  - [x] `sky launch -c min --cloud aws --cpus 2 --num-nodes 2 echo hi -i 0; sky status -r`
  - [x] `sky launch -c min --cloud aws --cpus 2 echo hi; manually terminate the VM; sky status -r`
  - [x] `sky launch --cloud gcp --cpus 2 --num-nodes 2 echo hi -i 0; sky status -r`
  - [x] `sky launch --cloud azure --cpus 2 --num-nodes 2 echo hi -i 0; sky status -r`
  - [x] `sky launch --cloud azure --cpus 2 echo hi -i 0; sky status -r`
  - [x] `sky launch --cloud ibm --cpus 2 --num-nodes 2 echo hi -i 0; sky status -r`
  - [x] `sky launch --cloud oci --cpus 2 --num-nodes 2 echo hi -i 0; sky status -r`
  - [x] `sky launch --cloud lambda --cpus 2+ --num-nodes 2 echo hi -i 0 --down; sky status -r`
  - [ ] `sky launch --cloud scp --cpus 2 echo hi -i 0; sky status -r` (not able to launch a cluster due to the test account failure)
- [x] All smoke tests: `pytest tests/test_smoke.py`
- [x] All smoke tests: `pytest tests/test_smoke.py --aws`
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/status_lib.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+"""Statuses enum for SkyPilot resources."""
```
Comment (Michaelvll) [inline sky/status_lib.py:1]: Moved the Status enum out of `global_user_state` to avoid cyclic imports.

File: sky/backends/backend_utils.py [hunk 2]
```diff
@@ -2010,14 +1662,42 @@ def check_owner_identity(cluster_name: str) -> None:
...
+        # Do not use refresh cluster status during teardown, as that will
+        # cause inifinite recursion by calling cluster status refresh
+        # again.
+        # Post teardown cleanup be done later in this function, which will remove
+        # the cluster entry from the status table & the ssh config file.
+        backend.teardown_no_lock(handle,
+                                 terminate=True,
+                                 purge=False,
+                                 post_teardown_cleanup=False,
+                                 refresh_cluster_status=False)
```
Comment (romilbhardwaj) [inline sky/backends/backend_utils.py:1699]: While we're doing this refactor, should we try to move this functionality (terminating preempted TPU VMs) a little up in the stack? This method should likely only query the cloud API and this TPU-specific handling can perhaps be done in `_update_cluster_status_no_lock`, where we have other TPU-specific code too.

File: sky/clouds/cloud.py [hunk 3]
```diff
@@ -467,5 +468,11 @@ def check_disk_tier_enabled(cls, instance_type: str,
         """
         raise NotImplementedError
 
+    @classmethod
+    def query_status(cls, name: str, tag_filters: Dict[str, str],
+                     region: Optional[str], zone: Optional[str],
+                     **kwargs) -> List['status_lib.ClusterStatus']:
```
Comment (romilbhardwaj) [inline sky/clouds/cloud.py:474]: Can we also add a short docstring here very briefly explaining what this method is supposed to do and what it returns? I was implementing this for Kubernetes last week and took some time to understand what needs to be done to implement this : )

File: tests/test_yamls/pipeline_aws.yaml [hunk 4]
```diff
@@ -13,6 +13,7 @@ run: |
   echo SKYPILOT_TASK_ID: $SKYPILOT_TASK_ID
   echo SKYPILOT_TASK_IDS
   echo "$SKYPILOT_TASK_IDS"
+  sleep 10000
```
Comment (romilbhardwaj) [inline tests/test_yamls/pipeline_aws.yaml:16]: unintended? Or should this also go in in this PR?

**Review by romilbhardwaj — APPROVED** [review]
Thanks @Michaelvll! This makes it much cleaner to add a new cloud. 

Took a quick look at the code and tested with AWS and GCP. Looks great!

File: sky/clouds/cloud.py (hunk 3 above)
Comment (Michaelvll) [reply sky/clouds/cloud.py:474]: Good point! Just added. Thanks!

File: sky/backends/backend_utils.py (hunk 2 above)
Comment (Michaelvll) [reply sky/backends/backend_utils.py]: Actually, I think this should be handled within the `GCP.query_status()` function, once we have refactored the termination code path into the cloud class. Reason: the TPU VM will stay in PREEMPTED status and holds the quota, even if that cluster is completely useless and should be cleaned up immediately. 
I tend to leave it here, and once we refactored the termination code path, we can move this into the `GCP.query_status()`. Added acomment about this. Wdyt?

File: tests/test_yamls/pipeline_aws.yaml (hunk 4 above)
Comment (Michaelvll) [reply tests/test_yamls/pipeline_aws.yaml:16]: Yes, this is required to make the `pytest tests/test_smoke.py --aws` to pass. It was a mistake from a previous PR. The `pipeline_gcp.yaml` has this `sleep 10000`, but we missed it for this file.

File: sky/clouds/cloud.py [hunk 5]
```diff
@@ -472,6 +472,17 @@ def check_disk_tier_enabled(cls, instance_type: str,
     def query_status(cls, name: str, tag_filters: Dict[str, str],
                      region: Optional[str], zone: Optional[str],
                      **kwargs) -> List['status_lib.ClusterStatus']:
+        """Queries the latest status of the cluster from the cloud.
+
+        The global_user_state caches the status of the clusters, but the
+        actualy status of the clusters may change on the cloud, e.g., the
```
Comment (romilbhardwaj) [inline sky/clouds/cloud.py:478]: typo: *actual

**Review by romilbhardwaj — APPROVED** [review]
Awesome, thanks @Michaelvll!

**Comment by Michaelvll** [comment]
Tested (822ae9a):
- [x] `pytest tests/test_smoke.py`

**Pull request merged**